### DISCUSSION TITLE
How effective is firejail or apparmor at restricting root power?

### DISCUSSION THREAD

**Discussion by crocket** [netblue30/firejail#4947] [Q&A] [answered]
Can firejail restrict root? Can apparmor? How effectively can they restrict root privilege?

**Comment by topimiettinen** [comment, marked as answer]
Firejail uses Linux namespaces (mount, network, PID), seccomp and capabilities to restrict sandboxed processes. They can be applied to any UID, also 0 (root). Compromised privileged process may still be able to use the remaining privileges to gain others. For example, capability CAP_SYS_ADMIN gives a lot of privileges but if it's taken away, a compromised process might still use CAP_FSETID to create a SUID file, use CAP_SYS_RAWIO to write directly to `/dev/sda`, CAP_SYS_MODULE to load malicious modules etc. Since file system binaries are owned by UID 0, it may be possible to overwrite `/sbin/init` even without any capabilities and seccomp can't help there either. Mount namespaces can be avoided by making device nodes and mounting file systems without restrictions somewhere. Still, Firejail isn't useless at all, attacks aren't always so clever and there may be enough easier targets if simple attacks fail somewhere.

MAC systems (AppArmor, SELinux, Smack, TOMOYO) can make further restrictions on top of what Firejail provides. Rules for a daemon could only allow it to read its own configuration, write to its own logs and state files, but not for example modify state files of another daemon even if both run as UID 0, or overwrite `/sbin/init` even if it's owned by UID 0. Of course, such rules must exist, otherwise there could be no protection. For example AppArmor has pretty small set of community provided rules and there's nothing similar for TOMOYO. Even SELinux doesn't protect all daemons.

But if the attackers have 0days for the kernel (or they use CPU weaknesses) none of the tricks above may help. In that case even non-zero UID might not help either. Insecure configuration is a great source of vulnerabilities, but also fixable. Firewalls could block some attacks. If a demon doesn't listen to any external network interfaces, it may be hard to attack, at least directly from the network.

**Reply by crocket** [reply]
Okay, I will not try to sandbox root or highly privileged capabilities with firejail.

**Reply by crocket** [reply]
By the way, is it possible to restrict dnsmasq and dnscrypt-proxy with firejail if they are run as root? Firejail has dnsmasq.profile and dnscrypt-proxy.profile.
